### PR TITLE
refactor(sdk): simplify placeholders and logic

### PR DESCRIPTION
**Pull request by connor-mccarthy**
**Description of your changes:**
This PR simplifies the implementation of in-memory placeholders, as well as simplifies/removes logic rendered unnecesary by the changes to placeholders.

This change is primarily driven the realization that we do not need to deserialize placeholder strings from IR into the placeholder objects that are used for serialization, as they do not offer us anything after the initial serialization due to the immutability of loaded PipelineSpec.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. 
[Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/components/pipeline_task.py [hunk 1]
```diff
@@ -208,141 +206,25 @@ def dependent_tasks(self) -> List[str]:
...
-            input_name: input_spec
-            for input_name, input_spec in component_inputs.items()
-        }
-        component_outputs = component_spec.outputs or {}
-        outputs_dict = {
-            output_name: output_spec
-            for output_name, output_spec in component_outputs.items()
-        }
-
-        def check_input_type_and_convert_to_placeholder(
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/pipeline_task.py:236]: For reviewer:

Much of this logic has nothing to do with the arguments provided.

The logic that _is not_ related to the arguments is moved to `structures.py`.

The logic that _is_ related to the arguments is either (a) fully captured by our type checking logic or (b) can be handled by `placeholders.convert_command_line_element_to_string`, reducing code duplication.

File: sdk/python/kfp/components/structures.py [hunk 2]
```diff
@@ -313,22 +313,11 @@ def from_container_dict(
         Returns:
             ContainerSpecImplementation: The ContainerSpecImplementation instance.
         """
-        args = container_dict.get('args')
-        if args is not None:
-            args = [
-                placeholders.maybe_convert_placeholder_string_to_placeholder(
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/structures.py:319]: For reviewer:

This logic is redundant, as conversion is handled later in the compilation process. We do not need to eagerly handle this here.

File: sdk/python/kfp/components/placeholders.py [hunk 3]
```diff
@@ -15,277 +15,110 @@
...
 from typing import Any, Dict, List, Optional, Union
 
 from kfp.components import base_model
 from kfp.components import utils
 from kfp.components.types import type_utils
 
 
-class Placeholder(abc.ABC):
-    """Abstract base class for Placeholders.
+class Placeholder(abc.ABC, base_model.BaseModel):
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/placeholders.py:26]: For reviewer:

Subclass `BaseModel` to bypass the `BaseModel`'s logic on which types its attributes can be. This is a workaround and would be good to clean up in another PR.

File: sdk/python/kfp/compiler/compiler_test.py [hunk 4]
```diff
@@ -137,59 +137,6 @@ def flip_coin_graph_component():
                 with dsl.Condition(flip.output == 'heads'):
                     flip_coin_graph_component()
 
-    def test_compile_pipeline_with_misused_inputvalue_should_raise_error(self):
```
Comment (connor-mccarthy) [inline sdk/python/kfp/compiler/compiler_test.py:140]: Moved these tests to `PipelineSpecBuilder`.

File: sdk/python/kfp/components/pipeline_task.py [hunk 5]
```diff
@@ -570,3 +468,56 @@ def my_pipeline():
                 )
             self._task_spec.dependent_tasks.append(task.name)
         return self
+
+
+# TODO: this function should ideally be in the function kfp.components.structures.check_placeholder_references_valid_io_name, which does something similar, but this causes the exception to be raised at component definition time, rather than compile time. This would break tests that load v1 component YAML, even though that YAML is invalid.
```
Comment (chensun) [inline sdk/python/kfp/components/pipeline_task.py:473]: > but this causes the exception to be raised at component definition time, rather than compile time. This would break tests that load v1 component YAML, even though that YAML is invalid.

Should we change the test to assertRaises on component definition then?

File: sdk/python/kfp/components/placeholders.py [hunk 6]
```diff
@@ -402,150 +186,76 @@ def container_with_if_placeholder(output_path: OutputPath(str),
...
 
 
-def json_load_nested_placeholder_aware(
-    placeholder_string: str
-) -> Union[str, Dict[str, Union[str, List[str], dict]]]:
-    try:
-        return json.loads(placeholder_string, cls=CustomizedDecoder)
-    except json.JSONDecodeError:
-        return placeholder_string
+CONTAINER_PLACEHOLDERS = (IfPresentPlaceholder, ConcatPlaceholder)
```
Comment (chensun) [inline sdk/python/kfp/components/placeholders.py:228]: nit: prefix with underscore since you don't reference them outside the module. 

https://google.github.io/styleguide/pyguide.html#3162-naming-conventions

File: sdk/python/kfp/components/structures_test.py [hunk 7]
```diff
@@ -871,13 +865,10 @@ def test_concat_placeholder(self):
...
-                        placeholders.ConcatPlaceholder(items=[
-                            placeholders.InputValuePlaceholder(
-                                input_name='input1'),
-                            placeholders.InputValuePlaceholder(
-                                input_name='input2')
-                        ])
+                        'sh',
+                        '-c',
+                        'echo "$0"',
+                        "{{$.inputs.parameters['input1']}}+{{$.inputs.parameters['input2']}}",
```
Comment (chensun) [inline sdk/python/kfp/components/structures_test.py:871]: It looks like we had a bug with `ContactPlaceholder` that it added the "+" when it shouldn't be.
Do we have a serialization test that verifies it's no longer the case?

**Review by chensun — COMMENTED** [review]
Overall LGTM, just a few nitpicks and questions.

File: sdk/python/kfp/components/pipeline_task.py (hunk 5 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/components/pipeline_task.py:473]: I think it's possible my comment is unclear. Let me try again and let me know if this helps:

The validation logic (not this exact function) lived in  `PipelineTask.__init__` before this PR. However, while working on this placeholder refactoring, I noticed that the validation can be fully completed using only the component definition (it does not require knowing what the component arguments are). It therefore ought to live `component_factory.create_component_from_func`.

While working on this PR, I attempted to move it there, but this ended up breaking the `kubeflow-pipelines-component-yaml` test suite, since it caught errored(?) v1 component YAML on load, rather than instantiation as a task.

> Should we change the test to assertRaises on component definition then?

To avoid expanding the scope of this PR, I moved the logic (albeit refactored for clarity) back to its original location `PipelineTask.__init__`, maintaining the current state. So, the `assertRaises` still needs to be on the instantiation. I left a `TODO` to indicate that this should be dealt with, as I am opting not to within this PR.

Does that make sense? Can you let me know if you have any suggestions for how the comment can be clearer?

File: sdk/python/kfp/components/structures_test.py (hunk 7 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/components/structures_test.py:871]: I think [this test](https://github.com/kubeflow/pipelines/blob/7adcc831035473f7aa35ab661bab8c944b61754f/sdk/python/kfp/components/placeholders_test.py#L143-L164) is sufficient. Can you let me know if you think there is some aspect this doesn't capture?

File: sdk/python/kfp/components/placeholders.py (hunk 6 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/components/placeholders.py:228]: Updated

File: sdk/python/kfp/components/structures_test.py (hunk 7 above)
Comment (chensun) [reply sdk/python/kfp/components/structures_test.py:871]: Sorry, I think the error was previously on the deserialization path--which no longer exists after this change. So we should be good.

Looking at the diff here, previously `"{{$.inputs.parameters['input1']}}+{{$.inputs.parameters['input2']}}"` was parsed into 
```
placeholders.ConcatPlaceholder(items=[
                            placeholders.InputValuePlaceholder(
                                input_name='input1'),
                            placeholders.InputValuePlaceholder(
                                input_name='input2')
                        ])
```

which serializes into 

```
{"Concat": ["{{$.inputs.parameters['input1']}}", "{{$.inputs.parameters['input2']}}"]}
``` 

and resolved at runtime as 

```
"{{$.inputs.parameters['input1']}}{{$.inputs.parameters['input2']}}"
```

(no plus sign in the middle)

File: sdk/python/kfp/components/pipeline_task.py (hunk 5 above)
Comment (chensun) [reply sdk/python/kfp/components/pipeline_task.py:473]: Make sense, and it's fine by me to keep it here for now.

**Review by chensun — APPROVED** [review]
/lgtm
/approve
